lib/src/game/viewport.dart

```diff
@@ -0,0 +1,185 @@
+abstract class Viewport {
+  /// This configures the viewport with a new raw size.
+  /// It should immediately affect [getEffectiveSize] and [getRawSize].
+  /// This must be called by the engine at startup and also whenever the
+  /// size changes.
+  void resize(Vector2 newRawSize);
+
+  /// This transforms the canvas so that the coordinate system is viewport-
+  /// -aware. All your rendering logic should be put inside the lambda.
+  void render(Canvas c, void Function(Canvas c) renderGame);
+
+  /// This returns the effective size, after viewport transformation.
+  /// This is not the game widget size but for all intents and purposes,
+  /// inside your game, this size should be used as the real one.
+  Vector2 getEffectiveSize();
+
+  /// This returns the real widget size (well actually the logical Flutter
+  /// size of your widget). This is the raw size as it would be without any
+  /// viewport.
+  ///
+  /// You probably don't need to care about this if you are using a viewport.
+  Vector2 getRawSize();
+}
+
+/// This is the default viewport if you want no transformation.
+/// The rawSize is just propagated to the effective size and no
+/// translation is applied.
+/// This basically no-ops the viewport.
+class DefaultViewport extends Viewport {
+  Vector2 _size;
+
+  @override
+  void render(Canvas c, void Function(Canvas c) renderGame) {
+    renderGame(c);
+  }
+
+  @override
+  void resize(Vector2 newRawSize) {
+    _size = newRawSize;
+  }
+
+  @override
+  Vector2 getEffectiveSize() => _size;
+
+  @override
+  Vector2 getRawSize() => _size;
+}
+
+/// This is the most common viewport if you want to have full control of what
+/// the game looks like. Basically this viewport makes sure the ratio between
+/// width and height is *always* the same in your game, no matter the platform.
+///
+/// To accomplish this you choose a virtual size that will always match the
+/// effective size.
+///
+/// Under the hood, the Viewport will try to expand (or contract) the virtual
+/// size so that it fits the most of the screen as it can. So for example,
+/// if the viewport happens to be the same ratio of the screen, it will resize
+/// to fit 100%. But if they are different ratios, it will resize the most it
+/// can and then will add black (color is configurable) borders.
+///
+/// Then, inside your game, you can always assume the game size is the fixed
+/// dimension that you provided.
+///
+/// Normally you can pick a virtual size that has the same ratio as the most
+/// used device for your game (like a pretty standard mobile ratio if you
+/// are doing a mobile game) and then in most cases this will apply no
+/// transformation whatsoever, and if the a device with a different ratio is
+/// used it will try to adapt the best as possible.
+class FixedRatioViewport extends Viewport {
+  Paint borderColor;
+
+  Vector2 rawSize;
+  Vector2 scaledSize;
+  Vector2 resizeOffset;
+  Vector2 fixedSize;
+  double scale;
+
+  FixedRatioViewport(
```

Comment:
Suggestion: `FixedResolutionViewport`. Then how the ratio is handled could be defined by optional parameters, but this would be for the future, since we only supports fixed ratio

Reply:
I'm not sure if I understand the difference, ratio and resolution is the same thing in this context?

Reply:
I think that the strong thing on this viewport, is the fixed resolution, not the ratio, like when I am making a game for gameboy I need a `160×144` resolution. If the viewport will respect the aspect ratio or not is another matter

Reply:
if it respects the resolution it necessarily respects the ratio. but I will change to resolution, just bear in mind it will scale up as much as possible.